ValidateTokenAsync - New Path: Refactor result types

iNinja opened the pull request:
# ValidateTokenAsync - New Path: Refactor result types
<!-- Thank you for submitting a pull request to our repo. -->

<!-- If this is your first PR in the IdentityModel repo, please run through the checklist
below to ensure a smooth review and merge process for your PR. -->

- [x] You've read the [Contributor Guide](https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/blob/dev/Contributing.md) and [Code of Conduct](https://opensource.microsoft.com/codeofconduct/).
- [x] You've included unit or integration tests for your change, where applicable.
- [x] You've included inline docs for your change, where applicable.
- [x] If any gains or losses in performance are possible, you've included benchmarks for your changes. [More info](https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/wiki/Benchmarking-in-Wilson)
- [x] There's an open issue for the PR that you are making. If you'd like to propose a new feature or change, please open an issue to discuss the change or find an existing issue.

<!-- Once all that is done, you're ready to go. Open the PR with the content below. -->

Refactored Result types to reduce allocations and improve performance.

## Description
- Added Result type (before making this public we may want to rename it to avoid clashes with a future .NET result type)
- Removed all token validation result types (IssuerValidationResult, AudienceValidationResult, ...)
- Repurposed ValidationResult as a new result type for the new path.
- Added extra stack frames when returning validation results. Added basic caching for a single path to validate.
- Return `Result<ValidationResult, ExceptionDetail>` from `TokenValidationAsync` to limit `ValidationResult` to **only** success scenarios, simplifying scenarios like creating and accessing the ClaimsIdentity.
- Removed ValidateTokenAsyncInternal since we now return the same Result from all methods, it is no longer required.
- Added ValidationFailureType to ExceptionDetail
- Updated tests

## Next steps after this:

- Possibly rename it to something more generic for an Error type that can create exceptions.
- Add capabilities to ExceptionDetail to pass extra data to the exception when being created (invalid issuer, audience)
- Re-implement Activator to create the exception from Details in a AOT compliant way (was removed time ago to restore AOT compatibility)

## Benchmarks for this work:

| Method                                                                        | Mean      | Error    | StdDev   | P90       | P95       | P100      | Ratio | RatioSD | Gen0   | Gen1   | Allocated | Alloc Ratio |
|------------------------------------------------------------------------------ |----------:|---------:|---------:|----------:|----------:|----------:|------:|--------:|-------:|-------:|----------:|------------:|
| JsonWebTokenHandler_ValidateTokenAsyncWithTVP_CreateClaims                    |  38.81 us | 0.130 us | 0.284 us |  39.17 us |  39.26 us |  39.85 us |  1.00 |    0.00 | 0.6104 |      - |  16.48 KB |        1.00 |
| **JsonWebTokenHandler_ValidateTokenAsyncWithVP_CreateClaims**                     |  38.13 us | 0.097 us | 0.209 us |  38.39 us |  38.45 us |  38.61 us |  0.98 |    0.01 | 0.6104 |      - |  16.02 KB |        0.97 |
|                                                                               |           |          |          |           |           |           |       |         |        |        |           |             |
| JsonWebTokenHandler_ValidateTokenAsyncWithTVP_SucceedOnFifthAttempt           | 102.50 us | 0.773 us | 1.698 us | 105.37 us | 106.85 us | 107.95 us |  1.00 |    0.00 | 1.4648 |      - |  37.13 KB |        1.00 |
| JsonWebTokenHandler_ValidateTokenAsyncWithTVPUsingClone_SucceedOnFifthAttempt | 103.07 us | 0.250 us | 0.544 us | 103.90 us | 104.00 us | 104.43 us |  1.01 |    0.02 | 1.5869 |      - |  38.89 KB |        1.05 |
| **JsonWebTokenHandler_ValidateTokenAsyncWithVP_SucceedOnFifthAttempt**            |  69.09 us | 0.266 us | 0.572 us |  69.73 us |  70.15 us |  70.46 us |  0.67 |    0.01 | 1.2207 |      - |   32.7 KB |        0.88 |
|                                                                               |           |          |          |           |           |           |       |         |        |        |           |             |
| JsonWebTokenHandler_ValidateTokenAsyncWithTVP_SucceedOnThirdAttempt           |  68.77 us | 0.186 us | 0.397 us |  69.33 us |  69.52 us |  69.95 us |  1.00 |    0.00 | 0.8545 |      - |  22.18 KB |        1.00 |
| JsonWebTokenHandler_ValidateTokenAsyncWithTVPUsingClone_SucceedOnThirdAttempt |  68.80 us | 0.132 us | 0.288 us |  69.16 us |  69.25 us |  69.80 us |  1.00 |    0.00 | 0.8545 |      - |  23.23 KB |        1.05 |
| **JsonWebTokenHandler_ValidateTokenAsyncWithVP_SucceedOnThirdAttempt**            |  51.37 us | 0.122 us | 0.268 us |  51.68 us |  51.70 us |  51.95 us |  0.75 |    0.00 | 0.7935 |      - |   19.7 KB |        0.89 |
|                                                                               |           |          |          |           |           |           |       |         |        |        |           |             |
| JsonWebTokenHandler_ValidateTokenAsyncWithTVP                                 |  34.25 us | 0.131 us | 0.279 us |  34.66 us |  34.80 us |  35.42 us |  1.00 |    0.00 | 0.2441 |      - |   7.23 KB |        1.00 |
| JsonWebTokenHandler_ValidateTokenAsyncWithTVPUsingModifiedClone               |  50.68 us | 0.085 us | 0.175 us |  50.91 us |  50.93 us |  51.00 us |  1.48 |    0.01 | 0.3662 |      - |   9.66 KB |        1.34 |
| **JsonWebTokenHandler_ValidateTokenAsyncWithVP**                                  |  33.80 us | 0.032 us | 0.070 us |  33.90 us |  33.92 us |  33.93 us |  0.99 |    0.01 | 0.2441 |      - |    6.7 KB |        0.93 |
| JwtSecurityTokenHandler_ValidateTokenAsync                                    |  41.49 us | 0.087 us | 0.184 us |  41.70 us |  41.81 us |  42.07 us |  1.21 |    0.01 | 1.0376 | 0.0610 |  25.58 KB |        3.54 |

Sample exception message + stack trace generated with the new stack frames:
```
IDX10500: Signature validation failed. No security keys were provided to validate the signature.
   at Microsoft.IdentityModel.JsonWebTokens.JsonWebTokenHandler.ValidateSignature(JsonWebToken jwtToken, ValidationParameters validationParameters, BaseConfiguration configuration, CallContext callContext) in C:\git\wilson\src\Microsoft.IdentityModel.JsonWebTokens\JsonWebTokenHandler.ValidateSignature.cs:line 99
   at Microsoft.IdentityModel.JsonWebTokens.JsonWebTokenHandler.ValidateJWSAsync(JsonWebToken jsonWebToken, ValidationParameters validationParameters, BaseConfiguration configuration, CallContext callContext, CancellationToken cancellationToken) in C:\git\wilson\src\Microsoft.IdentityModel.JsonWebTokens\JsonWebTokenHandler.ValidateToken.Internal.cs:line 339
   at Microsoft.IdentityModel.JsonWebTokens.JsonWebTokenHandler.ValidateTokenAsync(SecurityToken token, ValidationParameters validationParameters, CallContext callContext, CancellationToken cancellationToken) in C:\git\wilson\src\Microsoft.IdentityModel.JsonWebTokens\JsonWebTokenHandler.ValidateToken.Internal.cs:line 139
   at Microsoft.IdentityModel.JsonWebTokens.JsonWebTokenHandler.ValidateTokenAsync(String token, ValidationParameters validationParameters, CallContext callContext, CancellationToken cancellationToken) in C:\git\wilson\src\Microsoft.IdentityModel.JsonWebTokens\JsonWebTokenHandler.ValidateToken.Internal.cs:line 86
```

Part of #2711

Commit:
Adding benchmark for new ValidateTokenAsync model vs old.

FuPingFranco committed:
Update benchmark/Microsoft.IdentityModel.Benchmarks/ValidateTokenAsyncWithVPTests.cs

Co-authored-by: Keegan Caruso <Keegan.Caruso@microsoft.com>

Commit:
Updated class naming to be more explicit and added benchmark using clone

iNinja committed:
Added benchmarks to compare performance on multiple validation scenarios

iNinja committed:
Removed StackFrame from ValidationResult
Adding Result and Unit type. Experiment around allocations and performance.

iNinja committed:
Work in progress

iNinja committed:
Added project

iNinja committed:
Added missing await

iNinja committed:
Removed tags, added implicit initialisers for Result type. Updated Result return values to use implicit initialisation

iNinja committed:
Merge branch 'dev' into iinglese/refactor-results

iNinja committed:
Removed IdentityComparer methods relating to removed result types. Updated tests for Algorithm and Audience validation

iNinja committed:
Removed ITokenValidationError interface. Updated tests

iNinja committed:
Replaced TokenValidationError with ExceptionDetails. Added stack frames to failures, added initial cache experiment for stack frames

iNinja committed:
Removed optionality from CancellationToken parameter

iNinja committed:
Updated tests

iNinja committed:
Added ClaimsIdentity to ValidationResult

iNinja committed:
Merge branch 'dev' into iinglese/refactor-results

iNinja committed:
Updated benchmarks to match the result types and added CancellationToken

iNinja committed:
Removed test consoleapp, re-grouped benchmarks for better comparison

iNinja committed:
Removed unit type

iNinja committed:
Removed TokenValidationError since it is no longer used

iNinja committed:
Restored ExceptionType type name

iNinja committed:
Restored ExceptionFromType method in ExceptionDetail

jennyf19 commented on benchmark/Microsoft.IdentityModel.Benchmarks/ValidateTokenAsyncTests.cs:
```diff
@@ -98,7 +99,16 @@ public async Task<TokenValidationResult> JsonWebTokenHandler_ValidateTokenAsyncW
+        [BenchmarkCategory("ValidateTokenAsync_Success"), Benchmark]
+        public async Task<bool> JsonWebTokenHandler_ValidateTokenAsyncWithVP()
+        {
+            // Because ValidationResult is an internal type, we cannot return it in the benchmark.
```
do you want to open an issue to track this?

iNinja replied:
Added #2797

jennyf19 commented on src/Microsoft.IdentityModel.JsonWebTokens/JsonWebTokenHandler.DecryptToken.cs:
```diff
@@ -23,59 +24,49 @@ public partial class JsonWebTokenHandler : TokenHandler
+            (IList<SecurityKey>? contentEncryptionKeys, ExceptionDetail? exceptionDetail) result =
+                GetContentEncryptionKeys(jwtToken, validationParameters, configuration, callContext);
+
+            if (result.exceptionDetail != null)
```
can `result` be null?

iNinja replied:
Result is a value type and can therefore not be null. I did add an override for the empty initialisation to prevent the creation of an invalid Result object through extension.

jennyf19 commented on src/Microsoft.IdentityModel.JsonWebTokens/JsonWebTokenHandler.DecryptToken.cs:
```diff
@@ -97,14 +88,16 @@ internal TokenDecryptionResult DecryptToken(
                 {
-                    if (LogHelper.IsEnabled(EventLogLevel.Informational))
-                        LogHelper.LogInformation(TokenLogMessages.IDX10904, key);
+                    // TODO: Move to CallContext
```
no logic in the `if` now...remove?

iNinja replied:
Commented the rest of the `if` statement, to be removed in a coming PR.

jennyf19 commented on src/Microsoft.IdentityModel.JsonWebTokens/JsonWebTokenHandler.ValidateSignature.cs:
```diff
@@ -170,36 +173,36 @@ private static (SignatureValidationResult? validResult, bool KidMatched, KeyMatc
-                SignatureValidationResult result = ValidateSignatureWithKey(jwtToken, key, validationParameters, callContext);
-                if (result.IsValid)
+                Result<SecurityKey, ExceptionDetail> result = ValidateSignatureWithKey(jwtToken, key, validationParameters, callContext);
+                if (result.IsSuccess)
```
null check on `result`?

iNinja replied:
Value type explained in another comment.

jennyf19 commented on src/Microsoft.IdentityModel.JsonWebTokens/JsonWebTokenHandler.ValidateSignature.cs:
```diff
@@ -208,40 +211,36 @@ private static SignatureValidationResult ValidateSignatureWithKey(
-                    ValidationFailureType.SignatureValidationFailed,
-                    result.ExceptionDetail);
+
+            if (!result.IsSuccess)
```
null check?

iNinja replied:
Value type explained in another comment.

jennyf19 commented on src/Microsoft.IdentityModel.JsonWebTokens/JwtTokenUtilities.DecryptTokenResult.cs:
```diff
@@ -2,12 +2,12 @@
 using System.Text;
 using Microsoft.IdentityModel.Tokens;
-using Microsoft.IdentityModel.Logging;
+//using Microsoft.IdentityModel.Logging;
```
remove?

iNinja replied:
Removed.

jennyf19 commented on src/Microsoft.IdentityModel.JsonWebTokens/JwtTokenUtilities.DecryptTokenResult.cs:
```diff
@@ -62,8 +50,9 @@ internal static TokenDecryptionResult DecryptJwtToken(
                 {
-                    if (LogHelper.IsEnabled(EventLogLevel.Warning))
-                        LogHelper.LogWarning(TokenLogMessages.IDX10607, key);
+                    // TODO: Move to CallContext
```
will this happen in this PR or the next?

iNinja replied:
Next. We are moving most of our logs out by providing all the validated data as part of the result.
Need to figure it we still need to provide this information through the CallContext or if that can be dropped.

iNinja committed:
Override Result's empty initialiser and annotate it to prevent wrong initialisation

iNinja committed:
Removed commented code.

iNinja committed:
Commented empty if statement

iNinja committed:
Added cached stack frames and nullability annotations for ValidateTokenAsync
Changed return type to use Result and removed error information from ValidationResult
Added method to add StackFrames to ExceptionDetail

iNinja committed:
Added stack frames for ReadToken and DecryptToken

iNinja committed:
Added ValidationFailureType to ExceptionDetail

iNinja committed:
Updated tests to use ValidationFailureType

keegan-caruso commented on src/Microsoft.IdentityModel.Abstractions/Result.cs:
```diff
@@ -0,0 +1,159 @@
+    /// </summary>
+    /// <typeparam name="TResult"></typeparam>
+    /// <typeparam name="TError"></typeparam>
+    public readonly struct Result<TResult, TError> : IEquatable<Result<TResult, TError>>
```
I'm not sure we want a public type simply called `Result`

iNinja replied:
I am looking for an alternative name. If you can come up with a suggestion, I’ll happily take it.

Should be mentioned in the description, I believe.

brentschmaltz replied:
@keegan-caruso we will come up with a better name.

keegan-caruso commented on src/Microsoft.IdentityModel.Abstractions/Result.cs:
```diff
@@ -0,0 +1,159 @@
+        /// <summary>
+        /// Gets a value indicating whether the result is successful.
+        /// </summary>
+        readonly public bool IsSuccess { get; }
```
```suggestion
        public readonly bool IsSuccess { get; }
```

iNinja committed:
Update src/Microsoft.IdentityModel.Abstractions/Result.cs

Co-authored-by: Keegan Caruso <Keegan.Caruso@microsoft.com>

keegan-caruso commented on src/Microsoft.IdentityModel.LoggingExtensions/Microsoft.IdentityModel.LoggingExtensions.csproj:
```diff
@@ -8,7 +8,7 @@
     <PackageId>Microsoft.IdentityModel.LoggingExtensions</PackageId>
     <PackageTags>.NET;Windows;Authentication;Identity;Extensions;Logging</PackageTags>
-    <TargetFrameworks>netstandard2.0</TargetFrameworks>
+    <TargetFrameworks>netstandard2.0;net8.0</TargetFrameworks>
```
Can you help me understand this change?

brentschmaltz replied:
+1, i don't see a reason for this.

iNinja replied:
Creeped in during my experiments, I believe. I'll revert it.

Thanks for noticing it!

keegan-caruso commented on src/Microsoft.IdentityModel.Tokens/Validation/Results/ValidationResult.cs:
```diff
@@ -1,112 +1,167 @@
+            ValidationParameters validationParameters)
         {
-            ValidationFailureType = ValidationFailureType.ValidationNotEvaluated;
+            TokenHandler = tokenHandler ?? throw new ArgumentNullException("TokenHandler cannot be null.");
```
when ArgumentNullException has a single string param, it is supposed to be the parameter name.

keegan-caruso commented on src/Microsoft.IdentityModel.JsonWebTokens/JsonWebTokenHandler.ValidateSignature.cs:
```diff
@@ -143,19 +151,17 @@ private static SignatureValidationResult ValidateSignatureUsingAllKeys(
         }
 
-        private static (SignatureValidationResult? validResult, bool KidMatched, KeyMatchFailedResult? failedResult) ValidateUsingKeys(
+        private static (Result<SecurityKey, ExceptionDetail>? validResult, bool KidMatched, KeyMatchFailedResult? failedResult) ValidateUsingKeys(
```
Instead of the implicit value tuple, this would read a little better with an internal record struct, similar to what you have done elsewhere here.

iNinja replied:
I would love to give this another go once we settle the return types for all our validations.
If we go the route of making the result type less generic, we can create separate record structs for each validation.

iNinja replied:
It would cause us to duplicate quite a bit of functionality between all of them as well.

keegan-caruso commented on src/Microsoft.IdentityModel.Abstractions/Result.cs:
```diff
@@ -0,0 +1,159 @@
+        /// 
+        /// </summary>
+        /// <param name="obj"></param>
+        /// <returns></returns>
```
the overrides can just do `<inheritdoc/>`

iNinja replied:
I'll look more into the use of `<inheritdoc/>` as we update the type. Thank you

brentschmaltz commented on src/Microsoft.IdentityModel.Abstractions/Result.cs:
```diff
@@ -0,0 +1,159 @@
+    /// </summary>
+    /// <typeparam name="TResult"></typeparam>
+    /// <typeparam name="TError"></typeparam>
+    public readonly struct Result<TResult, TError> : IEquatable<Result<TResult, TError>>
```
Since this is not abstract, the class should not be in M.IM.Abstractions.
Need to move to M.IM.Tokens.

iNinja replied:
Moved

iNinja committed:
Moved Result to Tokens project, made it internal. Reverted TargetFrameworks change to LoggingExtensions

iNinja committed:
Addressed PR feedback around argument null exceptions

iNinja committed:
Merge branch 'dev' into iinglese/refactor-results

# Conflicts:
#	test/Microsoft.IdentityModel.Tokens.Tests/Validation/AsyncValidatorsTests.cs
#	test/Microsoft.IdentityModel.Tokens.Tests/Validation/ValidationParametersTests.cs

FuPingFranco commented on benchmark/Microsoft.IdentityModel.Benchmarks/ValidateTokenAsyncTests.cs:
```diff
@@ -5,9 +5,11 @@
 using System.Threading.Tasks;
 using BenchmarkDotNet.Attributes;
 using BenchmarkDotNet.Configs;
+using Microsoft.IdentityModel.Abstractions;
```
Not needed anymore `Results.cs` was move to I.M.Tokens.

FuPingFranco commented on src/Microsoft.IdentityModel.Tokens/Validation/Validators.Algorithm.cs:
```diff
@@ -37,7 +37,7 @@ public static partial class Validators
-        internal static AlgorithmValidationResult ValidateAlgorithm(
+        internal static Result<string, ExceptionDetail> ValidateAlgorithm(
             string algorithm,
             SecurityKey securityKey,
```
Do we still need SecurityKey and SecurityToken parameters for both the delegate and our default?

iNinja replied:
Addressed in another related comment.

FuPingFranco commented on src/Microsoft.IdentityModel.Tokens/Validation/Validators.Audience.cs:
```diff
@@ -43,56 +43,40 @@ public static partial class Validators
         /// <remarks>An EXACT match is required.</remarks>
 #pragma warning disable CA1801 // TODO: remove pragma disable once callContext is used for logging
-        internal static AudienceValidationResult ValidateAudience(IList<string> tokenAudiences, SecurityToken? securityToken, ValidationParameters validationParameters, CallContext callContext)
+        internal static Result<string, ExceptionDetail> ValidateAudience(IList<string> tokenAudiences, SecurityToken? securityToken, ValidationParameters validationParameters, CallContext callContext)
```
Do we still need SecurityToken parameter for both our delegate and default?

iNinja replied:
We can discuss that as follow up. This work was about the result types primarily.

FuPingFranco commented on src/Microsoft.IdentityModel.Tokens/Validation/Validators.Audience.cs:
```diff
@@ -43,56 +43,40 @@ public static partial class Validators
         /// <exception cref="SecurityTokenInvalidAudienceException">If none of the 'audiences' matched either <see cref="TokenValidationParameters.ValidAudience"/> or one of <see cref="TokenValidationParameters.ValidAudiences"/>.</exception>
```
Will you be updating the comments on this PR?

iNinja replied:
No. We are still talking about further changes to the return type.
I believe we should make it our final step once everything is set in stone to have a pass through all the documentation for the new path and update / align it.

FuPingFranco commented on src/Microsoft.IdentityModel.Tokens/Validation/Validators.IssuerSigningKey.cs:
```diff
@@ -44,7 +45,7 @@ public static partial class Validators
         /// <exception cref="ArgumentNullException"> if 'securityKey' is null and ValidateIssuerSigningKey is true.</exception>
         /// <exception cref="ArgumentNullException"> if 'securityToken' is null and ValidateIssuerSigningKey is true.</exception>
         /// <exception cref="ArgumentNullException"> if 'validationParameters' is null.</exception>
```
Will you be updating the comments on this PR?

iNinja replied:
Answered in a related comment.

FuPingFranco commented on test/Microsoft.IdentityModel.JsonWebTokens.Tests/JsonWebTokenHandler.DecryptTokenTests.cs:
```diff
@@ -2,8 +2,8 @@
 using System;
-using System.Diagnostics;
 using System.IdentityModel.Tokens.Jwt.Tests;
+using Microsoft.IdentityModel.Abstractions;
```
Not needed anymore.

FuPingFranco commented on test/Microsoft.IdentityModel.JsonWebTokens.Tests/JsonWebTokenHandler.ReadTokenTests.cs:
```diff
@@ -3,6 +3,7 @@
 
 using System;
 using System.IdentityModel.Tokens.Jwt.Tests;
+using Microsoft.IdentityModel.Abstractions;
```
Not needed anymore.

FuPingFranco commented on test/Microsoft.IdentityModel.JsonWebTokens.Tests/JsonWebTokenHandler.ValidateSignatureTests.cs:
```diff
@@ -3,7 +3,7 @@
 using System;
 using System.IdentityModel.Tokens.Jwt.Tests;
-using Microsoft.IdentityModel.JsonWebTokens.Results;
+using Microsoft.IdentityModel.Abstractions;
```
Not needed anymore.

FuPingFranco commented on test/Microsoft.IdentityModel.Tokens.Tests/Validation/AlgorithmValidationResultTests.cs:
```diff
@@ -1,11 +1,11 @@
 using Microsoft.IdentityModel.TestUtils;
 using Microsoft.IdentityModel.Logging;
 using Xunit;
+using Microsoft.IdentityModel.Abstractions;
```
Not needed anymore.

FuPingFranco commented on test/Microsoft.IdentityModel.Tokens.Tests/Validation/AudienceValidationResultTests.cs:
```diff
@@ -2,19 +2,19 @@
 using Microsoft.IdentityModel.Tokens.Json.Tests;
 using Xunit;
 using System.Collections.Generic;
+using Microsoft.IdentityModel.Abstractions;
```
Not needed anymore.

FuPingFranco commented on test/Microsoft.IdentityModel.Tokens.Tests/Validation/IssuerValidationResultTests.cs:
```diff
@@ -2,9 +2,9 @@
-using System.Diagnostics;
 using System.Threading;
 using System.Threading.Tasks;
+using Microsoft.IdentityModel.Abstractions;
```
Not needed anymore.

FuPingFranco commented on test/Microsoft.IdentityModel.Tokens.Tests/Validation/LifetimeValidationResultTests.cs:
```diff
@@ -2,7 +2,7 @@
 
 using System;
-using System.Diagnostics;
+using Microsoft.IdentityModel.Abstractions;
```
Not needed anymore.

FuPingFranco commented on test/Microsoft.IdentityModel.Tokens.Tests/Validation/ReplayValidationResultTests.cs:
```diff
@@ -2,7 +2,7 @@
 
 using System;
-using System.Diagnostics;
+using Microsoft.IdentityModel.Abstractions;
```
Not needed anymore.

FuPingFranco commented on test/Microsoft.IdentityModel.Tokens.Tests/Validation/SigningKeyValidationResultTests.cs:
```diff
@@ -2,8 +2,8 @@
 using System;
-using System.Diagnostics;
 using System.IdentityModel.Tokens.Jwt;
+using Microsoft.IdentityModel.Abstractions;
```
Not needed anymore.

FuPingFranco commented on test/Microsoft.IdentityModel.Tokens.Tests/Validation/SigningKeyValidationResultTests.cs:
```diff
@@ -50,11 +61,10 @@ public static TheoryData<SigningKeyValidationTheoryData> SigningKeyValidationTes
                         SecurityToken = new JwtSecurityToken(),
                         ValidationParameters = new ValidationParameters(),
-                        SigningKeyValidationResult = new SigningKeyValidationResult(KeyingMaterial.SymmetricSecurityKey2_256)
+                        Result = new ValidatedSigningKeyLifetime(null, null, utcNow)
```
This test is a bit flaky. Could we do something like this instead?

        private readonly DateTime utcNow;
        
        public SigningKeyValidationResultTests()
        {
            utcNow = DateTime.UtcNow;
        }

FuPingFranco replied:
Here's the failed run log:

Microsoft.IdentityModel.Tokens.Validation.Tests.SigningKeyValidationResultTests.SecurityKey(theoryData: Valid_SecurityTokenIsPresent, NoExceptionExpected) 
      Duration: 12.2 sec

      Message: 
Microsoft.IdentityModel.TestUtils.TestException : Microsoft.IdentityModel.Tokens.Validation.Tests.SigningKeyValidationResultTests.SigningKeyValidationResultTests : Valid_SecurityTokenIsPresent


dateTime1 != dateTime2. '8/27/2024 10:18:42 PM' != '8/27/2024 10:18:44 PM'.


      Stack Trace: 
TestUtilities.AssertFailIfErrors(String testId, List`1 errors) line 371
TestUtilities.AssertFailIfErrors(CompareContext context) line 352
SigningKeyValidationResultTests.SecurityKey(SigningKeyValidationTheoryData theoryData) line 48

iNinja replied:
We are creating the `utcNow` variable when the test starts. It is no different than creating it in the class as you suggest.
This is the same way we do in our lifetime validation tests, but because this validation is taking longer, the tests are running for some seconds and causing the flakiness.
What would solve this problem for good is to implement the DateTimeProvider that we discussed on Monday since we could ensure that the test receives always the same `utcNow` provided by our mock.

I'll increase our epsilon for now, until we address the root cause.

FuPingFranco commented on test/Microsoft.IdentityModel.Tokens.Tests/Validation/TokenTypeValidationResultTests.cs:
```diff
@@ -2,7 +2,7 @@
 
 using System;
-using System.Diagnostics;
+using Microsoft.IdentityModel.Abstractions;
```
Not needed anymore.

FuPingFranco reviewed (approved):
LGTM, just a couple of minor things like removing the additional imports, flaky test and a couple of questions.

iNinja committed:
Addressed PR comments. Removed unused imports. Increased epsilon for datetime comparison in tests.
